runtime/runtime-core/jvm/src/aws/smithy/kotlin/runtime/content/BigDecimalJVM.kt

```diff
@@ -4,33 +4,52 @@
  */
 package aws.smithy.kotlin.runtime.content
 
-public actual class BigDecimal actual constructor(public val value: String) :
+import java.math.BigDecimal as JvmBigDecimal
+
+public actual class BigDecimal private constructor(private val delegate: JvmBigDecimal) :
     Number(),
     Comparable<BigDecimal> {
-    private val delegate = java.math.BigDecimal(value)
 
-    public actual constructor(mantissa: BigInteger, exponent: Int) : this(
-        java.math.BigDecimal(
-            java.math.BigInteger(mantissa.toString()),
-            exponent,
-        ).toPlainString(),
-    )
+    private companion object {
+        /**
+         * Returns a new or existing [BigDecimal] wrapper for the given delegate [value]
+         * @param value The delegate value to wrap
+         * @param left A candidate wrapper which may already contain [value]
+         * @param right A candidate wrapper which may already contain [value]
+         */
+        fun coalesceOrCreate(value: JvmBigDecimal, left: BigDecimal, right: BigDecimal): BigDecimal = when (value) {
+            left.delegate -> left
+            right.delegate -> right
+            else -> BigDecimal(value)
+        }
+    }
+
+    public actual constructor(value: String) : this(JvmBigDecimal(value))
+    public actual constructor(mantissa: BigInteger, exponent: Int) : this(JvmBigDecimal(mantissa.delegate, exponent))
 
     public actual fun toPlainString(): String = delegate.toPlainString()
-    actual override fun toByte(): Byte = delegate.toByte()
-    actual override fun toDouble(): Double = delegate.toDouble()
-    actual override fun toFloat(): Float = delegate.toFloat()
-    actual override fun toInt(): Int = delegate.toInt()
-    actual override fun toLong(): Long = delegate.toLong()
-    actual override fun toShort(): Short = delegate.toShort()
+    public actual override fun toString(): String = delegate.toString()
+    public actual override fun toByte(): Byte = delegate.toByte()
+    public actual override fun toDouble(): Double = delegate.toDouble()
+    public actual override fun toFloat(): Float = delegate.toFloat()
+    public actual override fun toInt(): Int = delegate.toInt()
+    public actual override fun toLong(): Long = delegate.toLong()
+    public actual override fun toShort(): Short = delegate.toShort()
 
-    actual override fun equals(other: Any?): Boolean = other is BigDecimal && delegate == other.delegate
+    public actual override fun equals(other: Any?): Boolean = other is BigDecimal && delegate == other.delegate
+    public actual override fun hashCode(): Int = 31 + delegate.hashCode()
```

Comment:
Usually 31 is multiplied in these hashCode implementations, is this intentional addition? Is there any reason we can't just take the delegate's hash code?

Reply:
Yes, I intentionally used addition instead of multiplication although either should work. I wanted to avoid using the unmodified hash code from the delegate because that would bin them the same if they were both added to a hashed container (e.g., `Set`). That's unlikely but also trivial to avoid.